File: sdks/python/agentreel/redact.py
"""Redact secrets and PII from trajectory data."""

from __future__ import annotations

import re

_PATTERNS: list[tuple[re.Pattern, str]] = [
    # API keys (sk-...)
    (re.compile(r"(?i)(sk-[a-zA-Z0-9]{20,})"), "[REDACTED_API_KEY]"),
    # API key assignments
    (re.compile(r'(?i)(api[_-]?key\s*[:=]\s*)[\'"]?([a-zA-Z0-9_\-]{20,})[\'"]?'), r"\1[REDACTED]"),
    # Bearer tokens
    (re.compile(r"(?i)(bearer\s+)([a-zA-Z0-9_\-.]{20,})"), r"\1[REDACTED]"),
    # AWS keys
    (re.compile(r"AKIA[0-9A-Z]{16}"), "[REDACTED_AWS_KEY]"),
    # Passwords
    (re.compile(r'(?i)(password\s*[:=]\s*)[\'"]?([^\s\'"]{4,})[\'"]?'), r"\1[REDACTED]"),
    # Tokens
    (re.compile(r'(?i)(token\s*[:=]\s*)[\'"]?([a-zA-Z0-9_\-.]{20,})[\'"]?'), r"\1[REDACTED]"),
    # Email addresses
    (re.compile(r"[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}"), "[REDACTED_EMAIL]"),
]


def redact(text: str) -> str:
    """Redact secrets and PII from text."""
    result = text
    for pattern, replacement in _PATTERNS:
        result = pattern.sub(replacement, result)
    return result


def redact_trajectory_file(input_path: str, output_path: str | None = None) -> None:
    """Redact a trajectory JSON file."""
    with open(input_path) as f:
        content = f.read()

    redacted = redact(content)

    out = output_path or input_path
    with open(out, "w") as f:
        f.write(redacted)

File: sdks/typescript/src/redact.ts
import { readFileSync, writeFileSync } from "node:fs";

const PATTERNS: Array<[RegExp, string]> = [
  // API keys (sk-...)
  [/(?:sk-[a-zA-Z0-9]{20,})/gi, "[REDACTED_API_KEY]"],
  // API key assignments
  [/(?:api[_-]?key\s*[:=]\s*)['"]?([a-zA-Z0-9_\-]{20,})['"]?/gi, "$&".replace(/['"]?([a-zA-Z0-9_\-]{20,})['"]?/, "[REDACTED]")],
  // Bearer tokens
  [/(bearer\s+)([a-zA-Z0-9_\-.]{20,})/gi, "$1[REDACTED]"],
  // AWS keys
  [/AKIA[0-9A-Z]{16}/g, "[REDACTED_AWS_KEY]"],
  // Passwords
  [/(password\s*[:=]\s*)['"]?([^\s'"]{4,})['"]?/gi, "$1[REDACTED]"],
  // Tokens
  [/(token\s*[:=]\s*)['"]?([a-zA-Z0-9_\-.]{20,})['"]?/gi, "$1[REDACTED]"],
  // Email addresses
  [/[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}/g, "[REDACTED_EMAIL]"],
];

export function redact(text: string): string {
  let result = text;
  for (const [pattern, replacement] of PATTERNS) {
    result = result.replace(pattern, replacement);
  }
  return result;
}

export function redactTrajectoryFile(
  inputPath: string,
  outputPath?: string
): void {
  const content = readFileSync(inputPath, "utf-8");
  const redacted = redact(content);
  writeFileSync(outputPath ?? inputPath, redacted);
}
